pdp-engine/src/health.rs

```diff
@@ -0,0 +1,297 @@
+use url::Url;
+
+/// Helper function to check if PDP is healthy at the given URL
+pub async fn is_healthy(base_url: &Url, client: &Client) -> bool {
+    let url = match base_url.join("healthy") {
+        Ok(url) => url,
+        Err(e) => {
+            error!("Failed to create health URL: {}", e);
+            return false;
+        }
+    };
+
+    match client.get(url).send().await {
+        Ok(resp) => resp.status().is_success(),
+        Err(e) => {
+            debug!("Health check error: {}", e);
+            false
+        }
+    }
+}
+
+/// Sleep for the given duration or until cancellation occurs.
+/// Returns true if cancelled, false if sleep completed.
+pub async fn sleep_or_cancel(
+    duration: Duration,
+    token: &CancellationToken,
+    cancel_msg: &str,
+) -> bool {
+    tokio::select! {
+        _ = sleep(duration) => false,
+        _ = token.cancelled() => {
+            info!("{}", cancel_msg);
+            true
+        }
+    }
+}
+
+/// Helper function that waits for PDP to become healthy with timeout and cancellation support.
+/// Returns Ok(()) if PDP became healthy, or Err with the reason for failure (timeout or cancelled).
+pub async fn wait_for_healthy(
+    base_url: &Url,
+    client: &Client,
+    timeout: Duration,
+    token: &CancellationToken,
+) -> Result<(), PDPError> {
+    let start = std::time::Instant::now();
+    let check_interval = Duration::from_millis(500);
+
+    while start.elapsed() < timeout {
+        // Check if cancelled
+        if token.is_cancelled() {
+            return Err(PDPError::ShutdownTriggered);
+        }
+
+        // Check health
+        if is_healthy(base_url, client).await {
+            return Ok(());
+        }
+
+        // Wait before next check
+        if sleep_or_cancel(check_interval, token, "").await {
+            return Err(PDPError::ShutdownTriggered);
+        }
+    }
+
+    Err(PDPError::HealthCheckTimeout(
+        "PDP failed to become healthy within the timeout period".into(),
+    ))
+}
+
+/// Main health monitoring loop that checks PDP's health and handles recovery
+pub async fn run_health_monitor(
+    base_url: Url,
+    client: Client,
+    runner: Arc<PDPRunner>,
+    token: CancellationToken,
+    check_interval: Duration,
+    timeout: Duration,
+) {
+    info!("Health monitor started");
+
+    loop {
+        if token.is_cancelled() {
+            info!("Health monitor shutting down");
+            break;
+        }
+
+        if check_health_and_wait(&base_url, &client, &token, check_interval).await {
+            continue;
+        }
+
+        // PDP is unhealthy - attempt recovery
+        if let Some(()) =
+            try_recover_unhealthy_pdp(&base_url, &client, &runner, &token, timeout).await
+        {
+            continue;
+        }
+
+        // Whether recovery was successful or not, wait before next health check cycle
+        if sleep_or_cancel(check_interval, &token, "Health monitor shutting down").await {
```

Comment:
If recovery was ok you get here only in the next iteration, just making sure it's expected

Reply:
This code has been removed